packages/core/src/browser/tree/tree-widget.tsx

```diff
@@ -729,54 +729,113 @@ export class TreeWidget extends ReactWidget implements StatefulWidget {
+    protected decorateIcon(node: TreeNode, icon: React.ReactNode): React.ReactNode {
+        // icon can be null or undefined
+        if (!icon) {
+            return;
         }
-
         const overlayIcons: React.ReactNode[] = [];
-        new Map(this.getDecorationData(node, 'iconOverlay').reverse().filter(notEmpty)
-            .map(overlay => [overlay.position, overlay] as [TreeDecoration.IconOverlayPosition, TreeDecoration.IconOverlay | TreeDecoration.IconClassOverlay]))
-            .forEach((overlay, position) => {
-                const iconClasses = [TreeDecoration.Styles.DECORATOR_SIZE_CLASS, TreeDecoration.IconOverlayPosition.getStyle(position)];
-                const style = (color?: string) => color === undefined ? {} : { color };
-                if (overlay.background) {
-                    overlayIcons.push(<span key={node.id + 'bg'} className={this.getIconClass(overlay.background.shape, iconClasses)} style={style(overlay.background.color)}>
-                    </span>);
-                }
-                const overlayIcon = (overlay as TreeDecoration.IconOverlay).icon || (overlay as TreeDecoration.IconClassOverlay).iconClass;
-                overlayIcons.push(<span key={node.id} className={this.getIconClass(overlayIcon, iconClasses)} style={style(overlay.color)}></span>);
-            });
-
+        // if multiple overlays have the same overlay.position attribute, we'll de-duplicate those and only process the first one from the decoration array
+        const positionToOverlay = new Map<TreeDecoration.IconOverlayPosition, TreeDecoration.IconOverlay | TreeDecoration.IconClassOverlay>(
+            this.getDecorationData(node, 'iconOverlay').reverse().filter(notEmpty).map(overlay => [/* key */ overlay.position, /* value */ overlay])
+        );
+        for (const overlay of positionToOverlay.values()) {
+            const iconClasses = [TreeDecoration.Styles.DECORATOR_SIZE_CLASS, TreeDecoration.IconOverlayPosition.getStyle(overlay.position)];
+            const style = (color?: string) => color === undefined ? {} : { color };
+            if (overlay.background) {
+                overlayIcons.push(<span key={node.id + 'bg'} className={this.getIconClass(overlay.background.shape, iconClasses)} style={style(overlay.background.color)}></span>);
+            }
+            const overlayIcon = (overlay as TreeDecoration.IconOverlay).icon || (overlay as TreeDecoration.IconClassOverlay).iconClass;
+            overlayIcons.push(<span key={node.id} className={this.getIconClass(overlayIcon, iconClasses)} style={style(overlay.color)}></span>);
+        }
         if (overlayIcons.length > 0) {
             return <div className={TreeDecoration.Styles.ICON_WRAPPER_CLASS}>{icon}{overlayIcons}</div>;
         }
-
         return icon;
     }
 
+    /**
+     * Render the tree node tail decorations.
+     * @returns the generic class icon used for the tail decoration
+     */
+    protected getGenericClassIconDecoration(): string[] {
```

Comment:
The documentation is misleading, the method does not do any rendering, it only returns the generic icon decoration.

```suggestion
    /**
     * Get the generic icon decoration.
     * @returns the class icon used for generic decorations.
     */
    protected getGenericClassIconDecoration(): string[] {
```